torchtext/data/iterator.py

```diff
@@ -81,7 +111,17 @@ def data(self):
 
     def init_epoch(self):
         """Set up the batch generator for a new epoch."""
-        self.batches = batch(self.data(), self.batch_size, self.batch_size_fn)
+
+        if self._restored_from_state:
+            self.random_shuffler.random_state = self._random_state_this_epoch
+            self.batches = batch(self.data(), self.batch_size, self.batch_size_fn)
+            self._restored_from_state = False
+
+        else:
+            self._random_state_this_epoch = deepcopy(self.random_shuffler.random_state)
```

Comment:
I would avoid the deepcopy here by making `random_state` a `property` of the random shuffler whose "get" returns a copy of the state